crates/rattler_networking/src/authentication_storage/netrc_storage.rs

```diff
@@ -0,0 +1,89 @@
+//! Fallback storage for passwords.
+use netrc_rs::{Machine, Netrc};
+use serde::de::value;
+use std::{
+    collections::HashMap,
+    env,
+    io::Read,
+    path::PathBuf,
+    sync::{Arc, Mutex},
+};
+
+/// A struct that implements storage and access of authentication
+/// information backed by a on-disk JSON file
+#[derive(Clone)]
+pub struct NetRcStorage {
+    /// The path to the JSON file
+    pub path: PathBuf,
+
+    /// A mutex to ensure that only one thread accesses the file at a time
+    mutex: Arc<Mutex<()>>,
+
+    /// The netrc file contents
+    machines: HashMap<String, Machine>,
+}
+
+/// An error that can occur when accessing the fallback storage
+#[derive(thiserror::Error, Debug)]
+pub enum NetRcStorageError {
+    /// An IO error occurred when accessing the fallback storage
+    #[error("IO error: {0}")]
+    IOError(#[from] std::io::Error),
+    // /// An error occurred when (de)serializing the credentials
+    // #[error("JSON error: {0}")]
+    // JSONError(#[from] serde_json::Error),
+}
+
+impl NetRcStorage {
+    /// Create a new fallback storage with the given path
+    pub fn new() -> Self {
+        // Get the path to the netrc file
+        let path = match env::var("NETRC") {
+            Ok(val) => PathBuf::from(val),
+            Err(_) => match dirs::home_dir() {
+                Some(mut path) => {
+                    path.push(".netrc");
+                    path
+                }
+                None => PathBuf::from(".netrc"),
+            },
+        };
+
+        Self {
+            path: path.clone(),
+            mutex: Arc::new(Mutex::new(())),
+            machines: if path.exists() {
+                match std::fs::File::open(&path) {
```

Comment:
You could use std::fs::read_to_string and catch the io error to see if its kind is NotFound. That will maybe make this code shorter and more readable. 

You could also make it shorter by using some chaining but thats just  my preference 😉